### PR TITLE
[SCFD-2940] Transfer CHT solver validation

### PR DESCRIPTION
**Pull request by angranl-flex**
Transfer the `checkCHTSolverSettings()` in `Flow360Core/Scripts/flow360validation/validate/validateFlow360Case.py` into simulation framework.

### CODE REVIEW

File: flow360/component/simulation/simulation_params.py [hunk 1]
```diff
@@ -244,6 +245,12 @@ def check_consistency_wall_function_and_surface_output(cls, v):
         """Only allow wallFunctionMetric output field when there is a Wall model with a wall function enabled"""
         return _check_consistency_wall_function_and_surface_output(v)
 
+    @pd.model_validator(mode="after")
+    @classmethod
+    def check_cht_solver_settings(cls, v):
+        """Check the Conjugate Heat Transfer settings, transferred from checkCHTSolverSettings"""
+        return _check_cht_solver_settings(v)
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/simulation_params.py:252]: don't use single letter variables names
```suggestion
    def check_cht_solver_settings(cls, value):
        """Check the Conjugate Heat Transfer settings, transferred from checkCHTSolverSettings"""
        return _check_cht_solver_settings(value)
```

can you run pylint on this file? It should be caught by pylint

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 2]
```diff
@@ -3,9 +3,16 @@
 """
 
 from flow360.component.flow360_params.flow360_fields import get_aliases
-from flow360.component.simulation.models.surface_models import Wall
-from flow360.component.simulation.models.volume_models import Fluid
+from flow360.component.simulation.models.surface_models import (  # HeatFlux,; Temperature,
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:6]: don't leave commented code

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 3]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+
+    # for model in v.models:
+    #     if isinstance(model, Wall) and isinstance(model.heat_spec, Temperature):
+    #         raise ValueError(
+    #             "SolidIsothermalWall boundary is defined with no definition of volume zone of heat transfer."
+    #         )
+    #     if isinstance(model, Wall) and isinstance(model.heat_spec, HeatFlux):
+    #         raise ValueError(
+    #             "SolidAdiabaticWall boundary is defined with no definition of volume zone of heat transfer."
+    #         )
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:83]: don't leave commented code

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 4]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+
+    for model in v.models:
+        if isinstance(model, Fluid):
+            if (
+                model.navier_stokes_solver
+                and model.navier_stokes_solver.type_name != "Compressible"
+            ):
+                raise ValueError(
+                    "Conjugate heat transfer can not be used with incompressible flow solver."
+                )
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:115]: Use meaningful variable names. Probably here `v->params` will be better name

File: tests/simulation/params/test_validators_params.py [hunk 5]
```diff
@@ -97,3 +195,86 @@ def test_numerical_dissipation_mode_validator(
...
+    # message = (
+    #     "SolidAdiabaticWall boundary is defined with no definition of volume zone of heat transfer."
+    # )
+
+    # # Invalid simulation params
+    # with SI_unit_system, pytest.raises(ValueError, match=re.escape(message)):
+    #     _ = SimulationParams(
+    #         models=[fluid_model, wall_model_with_heatflux],
+    #         time_stepping=timestepping_unsteady,
+    #     )
```
Comment (maciej-flexcompute) [inline tests/simulation/params/test_validators_params.py:240]: remove

File: tests/simulation/params/test_validators_params.py [hunk 6]
```diff
@@ -97,3 +195,86 @@ def test_numerical_dissipation_mode_validator(
...
+    surface_output_with_residual_HeatSolver,
+    wall_model_with_temperature,
+    wall_model_with_heatflux,
+    fluid_model,
+    fluid_model_with_initial_condition,
+    solid_model,
+    solid_model_without_initial_condition,
+    solid_model_without_specific_heat_capacity,
+    timestepping_steady,
+    timestepping_unsteady,
```
Comment (maciej-flexcompute) [inline tests/simulation/params/test_validators_params.py:210]: if we don't reuse fixtures here, maybe better to define inside test function?

File: tests/simulation/params/test_validators_params.py [hunk 7]
```diff
@@ -29,6 +41,14 @@ def surface_output_with_numerical_dissipation():
     return surface_output
 
 
+@pytest.fixture()
+def surface_output_with_residual_HeatSolver():
```
Comment (maciej-flexcompute) [inline tests/simulation/params/test_validators_params.py:45]:
```suggestion
def surface_output_with_residual_heat_solver():
```

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 8]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
 
 
+def _check_cht_solver_settings(v):
+    has_heat_transfer = False
+
+    models = v.models
+
+    if models:
+        for model in models:
+            if isinstance(model, Solid) and model.entities is not None:
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:56]: Solid entities cannot be none

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 9]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+                has_heat_transfer = True
+            if isinstance(model, Wall) and model.heat_spec is not None:
+                has_heat_transfer = True
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:59]: Why?

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 10]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+            v = _validate_cht_has_heat_transfer(v)
+
+    return v
+
+
+def _validate_cht_no_heat_transfer(v):
+
+    # for model in v.models:
+    #     if isinstance(model, Solid) and model.heat_equation_solver:
+    #         raise ValueError("Heat equation solver activated with no zone definition.")
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:73]: These has been checked in line 56?

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 11]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+
+    # for model in v.models:
+    #     if isinstance(model, Wall) and isinstance(model.heat_spec, Temperature):
+    #         raise ValueError(
+    #             "SolidIsothermalWall boundary is defined with no definition of volume zone of heat transfer."
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:78]: This are needed.
Also the error message needs to change.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 12]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+    #         )
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:79]: These are no longer needed. We allow iso thermal wall but not "SolidIsothermalWall"

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 13]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+    #     if isinstance(model, Wall) and isinstance(model.heat_spec, HeatFlux):
+    #         raise ValueError(
+    #             "SolidAdiabaticWall boundary is defined with no definition of volume zone of heat transfer."
+    #         )
+
+    if v.outputs:
+        for output in v.outputs:
+            if not hasattr(output, "output_fields"):
+                continue
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:88]: why we add this?

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 14]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+            if output.output_type in [
+                "SurfaceOutput",
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:90]: Prefer isinstance() The discriminator string may change.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 15]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+                "VoluemeOutput",
+                "SliceOutput",
+                "ProbeOutput",
+                "IsosurfaceOutput",
+            ]:
+                if "residualHeatSolver" in output.output_fields.items:
+                    raise ValueError(
+                        "Heat equation output variables: residualHeatSolver is requested with"
+                        " no definition of volume zone of heat transfer."
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:99]: volume zone of heat transfer is referring to the old JSON struct. Let's update these accordingly so user clearly know what to change.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 16]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+def _validate_cht_has_heat_transfer(v):
+
+    for model in v.models:
+        if isinstance(model, Fluid):
+            if (
+                model.navier_stokes_solver
+                and model.navier_stokes_solver.type_name != "Compressible"
+            ):
+                raise ValueError(
+                    "Conjugate heat transfer can not be used with incompressible flow solver."
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:114]: Remove. We do not support incompressible solver in simulation framework.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 17]
```diff
@@ -39,6 +46,103 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+                )
+
+    time_stepping = v.time_stepping
+    if isinstance(time_stepping, Unsteady):
+        for model_solid in v.models:
+            if isinstance(model_solid, Solid):
+                if model_solid.material and model_solid.material.specific_heat_capacity is None:
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:121]: .material is impossible to be None

**Review by benflexcompute — COMMENTED** [review]
General comment: Let's try getting the error message as clear as possible.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 18]
```diff
@@ -39,6 +50,71 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+
+    if params.outputs:
+        for output in params.outputs:
+            if isinstance(
+                output, (SurfaceOutput, VolumeOutput, SliceOutput, ProbeOutput, IsosurfaceOutput)
+            ):
+                if "residualHeatSolver" in output.output_fields.items:
+                    raise ValueError(
+                        f"Heat equation output variables: residualHeatSolver is requested in {output.output_type} with"
+                        " no definition of solid models."
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:81]:
```suggestion
                        " no `Solid` model defined."
```

Comment (benflexcompute) [reply flow360/component/simulation/validation/validation_simulation_params.py:81]: Change other "solid" like this too.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 19]
```diff
@@ -39,6 +50,71 @@ def _check_consistency_wall_function_and_surface_output(v):
...
+
+    for model in params.models:
+        if isinstance(model, Fluid) and isinstance(
+            model.initial_condition, NavierStokesInitialCondition
+        ):
+            for model_solid in params.models:
+                if isinstance(model_solid, Solid) and model_solid.initial_condition is None:
+                    raise ValueError(
+                        "In solid model, the initial condition needs to be specified "
+                        "for initialization with expressions."
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:112]:
```suggestion
                        "when the `Fluid` model uses expression as initial condition."
```

File: tests/simulation/params/test_validators_params.py [hunk 20]
```diff
@@ -97,3 +105,92 @@ def test_numerical_dissipation_mode_validator(
...
+            outputs=[surface_output_with_residual_heat_solver],
+        )
+
+    message = "In solid model, the initial condition needs to be specified for initialization with expressions."
+
+    # Invalid simulation params
+    with SI_unit_system, pytest.raises(ValueError, match=re.escape(message)):
+        _ = SimulationParams(
+            models=[fluid_model_with_initial_condition, solid_model_without_initial_condition],
+            time_stepping=timestepping_steady,
```
Comment (benflexcompute) [inline tests/simulation/params/test_validators_params.py:194]: Just inline the code here.
```suggestion
            time_stepping=Steady(),
```

File: flow360/component/simulation/validation/validation_simulation_params.py (hunk 3 above)
Comment (angranl-flex) [reply flow360/component/simulation/validation/validation_simulation_params.py]: Removed.

File: flow360/component/simulation/validation/validation_simulation_params.py (hunk 4 above)
Comment (angranl-flex) [reply flow360/component/simulation/validation/validation_simulation_params.py]: Yes, I have changed the variable names to more meaningful ones.

File: tests/simulation/params/test_validators_params.py (hunk 7 above)
Comment (angranl-flex) [reply tests/simulation/params/test_validators_params.py]: Updated.

File: flow360/component/simulation/simulation_params.py (hunk 1 above)
Comment (angranl-flex) [reply flow360/component/simulation/simulation_params.py]: I ran pylint on this file before I submitted it. There are other functions that has single variable names but are not caught by pylint.

File: tests/simulation/params/test_validators_params.py (hunk 20 above)
Comment (angranl-flex) [reply tests/simulation/params/test_validators_params.py]: Updated.

File: flow360/component/simulation/validation/validation_simulation_params.py (hunk 18 above)
Comment (angranl-flex) [reply flow360/component/simulation/validation/validation_simulation_params.py]: Yes, all the other solid model are changed according to the comment.

File: flow360/component/simulation/validation/validation_simulation_params.py [hunk 21]
```diff
@@ -39,6 +50,71 @@ def _check_consistency_wall_function_and_surface_output(v):
     return v
 
 
+def _check_cht_solver_settings(params):
```
Comment (benflexcompute) [inline flow360/component/simulation/validation/validation_simulation_params.py:53]: Let's add a docstring here indicating what this function is trying to replace in the casePipeline validation (checkCHTSolverSettings()).

**Review by benflexcompute — APPROVED** [review]

**Pull request merged**